[DataGrid] Add Density selector

DanailH opened the pull request:
Closes #567 

I have added a new property on the DataGrid -> `density` that can be one of 3 options: `compact`, `standard` and `comfortable`. The default value is `standard`. In addition, the density property also works with the `rowHeight` and `headerHeight` in a way that if those are provided it will take them into account when adjusting the size of the grid.

Density options:

- For `compact` size -> the height of the rows is lowered by 1/3 of their original value.
- For `comfortable` size -> the height of the rows is increased by 1/3 of their original value.

All the logic is contained within the new `DensitySelector` component.

https://deploy-preview-606--material-ui-x.netlify.app/components/data-grid/rendering/#density

**TODO:**
- [x] Documentation is missing from the PR. Once we agree on the API I'll add the docs and remove the `draft` status from the PR.
- [x] Add unit tests.

<img width="390" alt="Capture d’écran 2020-12-03 à 17 52 41" src="https://user-images.githubusercontent.com/3165635/101061053-5efaf580-3590-11eb-90af-8b2e191583ed.png">

oliviertassinari commented on packages/grid/_modules_/grid/components/icons/index.tsx:
```diff
@@ -12,3 +12,21 @@ export const ArrowDownwardIcon = createSvgIcon(
+);
+
+export const ViewStreamIcon = createSvgIcon(
+  <React.Fragment>
+    <path d="M0 0h24v24H0z" fill="none" />
```
dead path

```suggestion
```

oliviertassinari commented on packages/grid/_modules_/grid/components/DensityPicker.tsx:
```diff
@@ -0,0 +1,152 @@
+          horizontal: 'center',
+        }}
+        anchorEl={anchorEl}
+        keepMounted
```
Probably better for performance as not frequently used.

```suggestion
```

oliviertassinari commented on packages/grid/_modules_/grid/components/DensityPicker.tsx:
```diff
@@ -0,0 +1,152 @@
+  return (
+    <div className="MuiDataGrid-toolbarElement">
+      <Button
+        aria-controls="customized-menu"
```
This is an id, it can't be unique

oliviertassinari reviewed (commented):
It looks like a great start. 

- I don't think that the state approach can scale: 1. the `options` are overriding by props when the grid rerender with different props and `rowHeight` can be changed after the first render.
- I think that the feature should be opt-in. We might start no to ask the developers to import and inject the component explicitly. 
- There is a new scrollbar on the header

<img width="843" alt="Capture d’écran 2020-11-19 à 16 57 42" src="https://user-images.githubusercontent.com/3165635/99690571-9645ae80-2a88-11eb-8829-8d54b6287c02.png">

- It could be interesting to use the icon on the button to signal the current state of the density.
- @dtassone DataGrid or XGrid?

dtassone commented on packages/grid/_modules_/grid/components/DensityPicker.tsx:
```diff
@@ -0,0 +1,152 @@
+    }
+  }, [setGridState, forceUpdate]);
+
+  const renderDensityOptions: Array<React.ReactElement> = DensityOptions.map((option, index) => (
```
We have 2 distincts components here.
The button, in the toolbar 
The menu that let you set the density
Then we should have a useDensityPicker to glue them using the apiRef.
So => button => apiRef.current.showDensityPicker => MenuDisplayed => 
apiRef.current.setDensity('normal') => setRowHeight(options.rowHeight * 1) 🤔 
apiRef.current.hideDensityPicker

This way the button is independant to the menu, and user can render the menu without the button or render their own component and use the api to change the density...
Then we should have an option to enableDensityPicker?

The flow for feature is Component calls apiRef that is on a dedicated hook, the hook update the state and trigger a rerender

DanailH replied:
I agree with how the functionality should he chained but IMO the button and the menu should go together. If you hide the button the menu should also be removed from the Grid and them the user has to implement both, I think that this is expected dev flow. What i can do is move the `Menu` config into its own component but that seems like an unnecessary abstraction since we are already using it from MUI.

oliviertassinari replied:
I would propose we start with the simplest possible option. It's still unclear if developers will use it. Or what they will need in terms of customization power. I think that deferring would be better.

dtassone commented:
> It looks like a great start.
> 
> * I don't think that the state approach can scale: 1. the `options` are overriding by props when the grid rerender with different props and 2. `rowHeight` can be changed after the first render.
> * I think that the feature should be opt-in. We might start now to ask the developers to import and inject the component explicitly.
> * There is a new scrollbar on the header
> 
> <img alt="Capture d’écran 2020-11-19 à 16 57 42" width="843" src="https://user-images.githubusercontent.com/3165635/99690571-9645ae80-2a88-11eb-8829-8d54b6287c02.png">
> 
> * It could be interesting to use the icon on the Density toolbar button to signal the current state of the density.
Yes I agree, the current density icon should be used in the toolbar
> * @dtassone DataGrid or XGrid?
XGrid

DanailH commented:
Ok I checked all the comments and did some additional tests to see how to rework the PR. I'll update the PR with all the requests in bulk. Regarding the state - I think we will need to keep what the current density is in the state. The reason for this is that because the the `rowHeight` is set via a prop and the density ultimately depends on the `rowHeight` value. 

There is another interesting question tho - if we go with the approach of having an API to change the density then we don't need to expose a prop for the `density` IMO. Either one or the other because if someone uses both we may and up with weird cases. I took the approach of providing a prop and @dtassone is suggesting an API so we need to decide (after that we will use the same approach for the other toolbar related features)

dtassone commented:
> Ok I checked all the comments and did some additional tests to see how to rework the PR. I'll update the PR with all the requests in bulk. Regarding the state - I think we will need to keep what the current density is in the state. The reason for this is that because the the `rowHeight` is set via a prop and the density ultimately depends on the `rowHeight` value.
> 
> There is another interesting question tho - if we go with the approach of having an API to change the density then we don't need to expose a prop for the `density` IMO. Either one or the other because if someone uses both we may and up with weird cases. I took the approach of providing a prop and @dtassone is suggesting an API so we need to decide (after that we will use the same approach for the other toolbar related features)

Things need to be built in a uniform manner. 
The apiRef drives the whole grid. It would be great to keep that pattern. Component => hook (update state) => component
Density is like all the other hooks like the sortModel, filters , selection... So you can set a prop, and change it internally with the apiRef, default prop discussion...

mbrookes commented on packages/grid/_modules_/grid/components/DensityPicker.tsx:
```diff
@@ -0,0 +1,152 @@
+  const [anchorEl, setAnchorEl] = React.useState(null);
+  const icons = useIcons();
+
+  const DensityShortIcon = icons!.densityShort!;
+  const DensityMediumIcon = icons!.densityMedium!;
+  const DensityTallIcon = icons!.densityTall!;
```
Density is normally low/medium/high (size is small/medium/large)

oliviertassinari replied:
<img width="534" alt="Capture d’écran 2020-11-23 à 20 54 58" src="https://user-images.githubusercontent.com/3165635/100008777-2ad44780-2dce-11eb-9f92-ab4dab1a3b36.png">

https://material-density.glitch.me/

vs.

https://material-ui.com/components/tables/#dense-table `<Table size="small">`

oliviertassinari replied:
What if we call the prop `size`?

mbrookes replied:
"density" is more accurate, but if consistency with the Table is important, then sure. (Is it a prop though?)

oliviertassinari replied:
Looking at https://next.material-ui.com/customization/density/#main-content, we use the two names interchangeably. I would be in favor of using `size` everywhere. I think that it will be simpler.

mbrookes replied:
It's mostly either `margin: dense`, or `size: small`, with one instance of `dense: true`, and one of `variant: dense`. 

Of the first two, `size: small` is closest, even if it doesn't quite describe the purpose.

DanailH replied:
Ok so I kept the prop name `density`, should I change it to `size` and also if yes then the same will apply to the button label right?

oliviertassinari replied:
I personally like the approach of using:

- "density" for the copywriting.
- `size?: "small" | "medium" | "large"` for the code API.

dtassone replied:
The hook is called `useDensity`, the component is `densityPicker` and the label in the toolbar is `density` so why would we name it size in the prop? 
We should call it density as it's less confusing for us as well as for users and also, it's easier to find and understand in the list of props. Button density in the toolbar matches the density prop, it's logical and straightforward.

Otherwise, we can rename everything as `useSize`, `SizePicker`, and `size` in the label, and have a size property.
It doesn't make sense to me to have something named `useFoo`, `Foo` and expose it as `Bar`

oliviertassinari replied:
> The hook is called useDensity, the component is densityPicker and the label in the toolbar is density so why would we name it size in the prop?

We can rename them all to "size".

I think that size is superior in the code because more generic and allow consistency. Why don't we call `<Button size="small">` and not `<Button density="dense">`?  https://ant.design/components/table/#components-table-demo-size, https://material-ui.com/components/tables/#dense-table.

DanailH commented:
Ok, it took a but of time but the PR is now ready to be reviewed. The biggest change is that I need to separate the props that the density affects so I created a `density` section in the state that holds the density size, `rowHeight` and `headerHeight` as these are changed when you change the density. 

In addition I also added the missing docs and unit tests.

PS: The unit tests that I added are failing because there is a warning in the console (only happens on the tests) that is 
```
Warning: findDOMNode is deprecated in StrictMode. findDOMNode was passed an instance of Transition which is inside StrictMode.
```

If someone has any ideas how to solve this I would really appreciate it.

dtassone commented on packages/grid/_modules_/grid/components/icons/index.tsx:
```diff
@@ -38,6 +38,21 @@ export const ColumnIcon = createSvgIcon(
 
 export const SeparatorIcon = createSvgIcon(<path d="M11 19V5h2v14z" />, 'Separator');
 
+export const ViewHeadlineIcon = createSvgIcon(
```
I think it would be better to rename those icons in density small, medium... As atm we don't understand how they are being used

oliviertassinari replied:
The convention is to use the name given in https://material.io/icons, as if the icon was imported from `@material-ui/icons`.

DanailH replied:
I figured it was like what @oliviertassinari that's why I named them like that. Also if they are named like this we can reuse then for other use cases later on if needed. If it is ok I'll resolve this conversation and keep their names as it is.

dtassone commented on packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts:
```diff
@@ -95,6 +97,7 @@ export const useStyles = makeStyles(
         '& .MuiDataGrid-colCell': {
           position: 'relative',
           display: 'flex',
+          alignItems: 'center',
```
have you checked all columns icons alignment with filters sort...
Col can be aligned to the left or right as well

DanailH replied:
I checked it - it seems ok. This is needed in order to keep the col cell separator line centred when the density changes.

dtassone commented on packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx:
```diff
@@ -0,0 +1,104 @@
+import { DensityOption } from '../../models/api/densityApi';
+import { densitySizeSelector } from '../../hooks/features/density';
+
+export const DensityPicker = React.memo(function DensityPicker() {
```
what if we want to render the menu from another button like in the column menu.
This way, it is all coupled together, so we can't

oliviertassinari replied:
> what if

Let's wait to get a request in this direction.

DanailH replied:
Yes that is the reason why I have it both in the same file. IMO it makes more sense to have them coupled together so if devs want to have their own component, when they disable the feature all presentational code related to it will be removed.

oliviertassinari replied:
Agree, I think that we should focus on the composability of the toolbar first.

dtassone commented on packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx:
```diff
@@ -0,0 +1,104 @@
+        Density
+      </Button>
+
+      <Menu
```
why not use the gridMenu? as mentioned before

oliviertassinari replied:
Damien has raised an interesting point, the `Menu` behaves like a modal, `GridMenu` is closer to what we want to refactor the Menu to behave like in v5.
So a proposal, we use `GridMenu` until v5 fixes it. *Ideally*, GridMenu should be swappable with the API of `Menu`.

DanailH replied:
Done

dtassone commented on packages/grid/_modules_/grid/hooks/features/density/densityState.ts:
```diff
@@ -0,0 +1,15 @@
+
+export interface DensityState {
+  size: Density;
+  rowHeight: number;
```
rowheight is already in options. Is it a duplicate?
we should only have one source of truth

dtassone commented on packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx:
```diff
@@ -0,0 +1,104 @@
+    }
+  }, [densitySize]);
+
+  const handleDensityPickerOpen = (event) => setAnchorEl(event.currentTarget);
```
all handled in the gridmenu...

dtassone commented on packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx:
```diff
@@ -0,0 +1,104 @@
+    apiRef!.current.setDensity(density, headerHeight, rowHeight);
+  }, [apiRef, density, rowHeight, headerHeight]);
+
+  const getSelectedDensityIcon = React.useCallback((): React.ReactElement => {
```
you could use a key value object and avoid the switch and the densityOptions block

DanailH replied:
I could but I also need to loop through the object to create the menu options and currently there is a problem with Typescript that if you have a variable that is of a custom type (from an Interface) then you need to add an iterator function to that interface handle the `for of` loop. I'm open to suggestions.

dtassone commented on packages/grid/_modules_/grid/hooks/features/density/densityState.ts:
```diff
@@ -0,0 +1,15 @@
+
+export interface DensityState {
+  size: Density;
+  rowHeight: number;
```
we have 2 sources of rowHeight now, it's a bit confusing 🤔

DanailH replied:
Yes I also don't like it much but the problem is 2 fold. One is the actual prop `rowHeight` which is set by devs and we should not control that one. The problem is that we do need a way to change its value in order to adjust the density. The idea behind this approach is that we never touch the props but rather if we need to do something with them we change their representation in the state.

dtassone commented on packages/grid/_modules_/grid/models/api/densityApi.ts:
```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
+import { DensityTypes, Density } from '../gridOptions';
+
+export interface DensityConfig {
```
where is this one used? Why not use the state?

DanailH replied:
You are right, its not use anywhere anymore. I'm removing it. Thanks for the catch ;)

dtassone commented on packages/grid/_modules_/grid/models/api/densityApi.ts:
```diff
@@ -0,0 +1,25 @@
+   * @param density
+   * @returns void
+   */
+  setDensity: (density: Density, headerHeight?, rowHeight?) => void;
```
ShowDensityMenu, HideDensityMenu

DanailH replied:
Do we really need API methods for those, It's just a toggle of a boolean. My argument is if people want to have their own custom density picker they will also most likely have their own menu and the functionality to open and close it.

IMO the part of the density picker that makes sense to be a public API is the way to change it since this is also the part of the functionality that affects the state and the grid in general.

oliviertassinari replied:
I'm personally in favor of removing all imperative APIs, and to only add them, on a case by case basis if asked by developers with a solid case.

oliviertassinari commented on packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx:
```diff
@@ -0,0 +1,104 @@
+      >
+        Density
+      </Button>
+
```
```suggestion
```

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+### Density Picker button
+
+The default way to change the density of the `DataGrid` is by using the `DensityPicker` component from the toolbar. In addition there is a property to hide the `DensityPicker` from the toolbar.
```
Components are in a global name space, `DensityPicker` is not OK, should be `GridDensityPicker`.

mbrookes replied:
```suggestion
The default way to change the density of the data grid is by using the `DensityPicker` component on the toolbar. The `disableDensityPicker` prop can be used to hide this control.
```

oliviertassinari commented on packages/storybook/src/stories/grid-toolbar.stories.tsx:
```diff
@@ -0,0 +1,27 @@
+    },
+  },
+};
+export const DensityPicker = () => {
```
Do we need this story? Maybe we should have 2 stories with `size="small"` and `size="large"`?

oliviertassinari replied:
In the documentation, I assume we will only display one variant of the size of the grid, which won't be enough for visual regression tests.

DanailH replied:
Makes sense. I'll add 2 stories and remove the default case.

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+{{"demo": "pages/components/data-grid/rendering/DensityPickerSmallGrid.js", "bg": "inline"}}
+
+### Density Picker button
```
Point 2. https://trello.com/c/1Q7hqXHu/2714-technical-documentation-guidelines

```suggestion
### Density picker button
```

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+### Density prop
+
+There are three possible density options to choose from: `small`, `medium` and `large`. By default the density of the `DataGrid` is set to `medium`. The default grid density can be changed by setting the `density` propery. The `density` propery works with relation to the `rowHeight` and `headerHeight` properties. This means that it will respect the values provided using `rowHeight` or `headerHeig ...
```
```suggestion
There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by setting the `density` prop. The `density` prop works with relation to the `rowHeight` and `headerHeight` properties. This means that it will respect the values provided using `rowHeight` or `headerHeight` props.
```

mbrookes replied:
Based off @oliviertassinari's:
```suggestion
There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by applying the `density` prop. The `density` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```

mbrookes commented on packages/grid/_modules_/grid/components/styled-wrappers/GridOverlay.tsx:
```diff
@@ -1,16 +1,21 @@
 import { classnames } from '../../utils';
-import { OptionsContext } from '../options-context';
+import { ApiContext } from '../api-context';
+// import { OptionsContext } from '../options-context';
```
Remove dead code?

mbrookes commented on packages/grid/_modules_/grid/components/styled-wrappers/GridOverlay.tsx:
```diff
@@ -1,16 +1,21 @@
 export function GridOverlay(props: GridOverlayProps) {
   const { className, style, ...other } = props;
-  const options = React.useContext(OptionsContext);
+  // const options = React.useContext(OptionsContext);
```
Remove dead code?

mbrookes commented on packages/grid/_modules_/grid/models/gridOptions.tsx:
```diff
@@ -49,6 +52,26 @@ export interface IconsOptions {
    * Icon displayed in between 2 column headers that allows to resize the column header.
```
```suggestion
   * Icon displayed in between two column headers that allows to resize the column header.
```
(Numbers up to ten are always spelled in full)

mbrookes commented on packages/grid/_modules_/grid/models/gridOptions.tsx:
```diff
@@ -49,6 +52,26 @@ export interface IconsOptions {
+}
+
+/**
+ * Available density types.
```
```suggestion
 * Available densities.
```

DanailH commented on packages/grid/_modules_/grid/models/gridOptions.tsx:
```diff
@@ -297,6 +325,10 @@ export interface GridOptions {
+  /**
+   * Set the density of the grid.
+   */
+  density: Density;
```
Done

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+### Size prop
+
+There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```
use the generic to not exclude XGrid.

```suggestion
There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the data grid is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```

mbrookes replied:
```suggestion
The vertical density data grid  can be set using the `size` prop. Regardless of its setting, the `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```

Avoid mixing size and density terminology, avoid duplicating info from the API docs (defaults, prop values). (I don't hold a strong opinion on this second point.

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+### Density picker button
+
+The default way to change the density of the data grid is by using the `DensityPicker` component from the toolbar. In addition, there is a prop to hide the `DensityPicker`.
```
This component is private, developers shouldn't know about it.

```suggestion
The default way to change the density of the data grid is by using the density picker component from the toolbar. In addition, there is a prop to hide the density picker.
```

mbrookes replied:
```suggestion
The default way for a user to change the density of the data grid is by using the density picker component from the toolbar. In addition, there is a prop to hide the density picker.
```

oliviertassinari commented on packages/grid/_modules_/grid/models/gridOptions.tsx:
```diff
@@ -46,9 +49,29 @@ export interface IconsOptions {
+   * Icon displayed in between two column headers that allows to resize the column header.
    */
   ColumnResize?: React.ElementType<{ className: string }>;
+
```
Missing descriptions x3

DanailH replied:
Sorry bout that, I've missed it, Done

oliviertassinari commented on packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx:
```diff
@@ -0,0 +1,100 @@
+  ));
+
+  return (
+    <div className="MuiDataGrid-toolbarElement">
```
```suggestion
    <React.Fragment>
```

oliviertassinari reviewed (commented):
Almost good to be shipped from my end.

dtassone commented on docs/src/pages/components/data-grid/rendering/DensityPickerSmallGrid.js:
```diff
@@ -0,0 +1,17 @@
+
+  return (
+    <div style={{ height: 300, width: '100%' }}>
+      <DataGrid {...data} size="small" />
```
at first sight, the prop name is bit weird and it does not feel linked to the density. Why not name it density?

oliviertassinari replied:
Already discussed in https://github.com/mui-org/material-ui-x/pull/606#discussion_r526992142.

dtassone commented:
Generic review
1. It seems that we are mainly using density but for some reason, we expose it as size which is quite confusing I think. To me, size is more a measure in pixel or any units... Compact, Normal, or small seems more like a density style, and in the code, you refer to it mainly as density so I think we should unify this vocabulary and keep one terminology.

2. I maintain my thing for splitting the components in 2, the button and the menu, all linked by the api.
 - Same way we develop the rest of the code
 - Give the flexibility to users to render the menu without the button, and outside the toolbar.
 - Allow opening the menu with api, and state

mbrookes commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -213,7 +213,7 @@ You can change the density of the rows and the column header.
 ### Size prop
 
-There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the `DataGrid` is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
+There are three possible density options to choose from: `small`, `medium`, and `large`. By default, the density of the data grid is set to `medium`. The default grid density can be changed by applying the `size` prop. The `size` prop will respect the values provided using the `rowHeight` or `headerHeight` props.
```
That's odd, I left a comment with a suggestion that fixed some of the issues with this description, while leaving some of it open for discussion. Could we please not just overrule in those cases?

mbrookes replied:
Thanks for adding the suggestion back @oliviertassinari. The comment was something to the effect that it isn't necessary to reproduce the API documentation in the description. This change also resolves the potential confusion around what the options actually apply to, and also the use of "default" to mean two different things.

DanailH replied:
I'm sorry about that @mbrookes I did merge it. I think git merges it weirdly if there are 2 suggestions on the same block.

mbrookes replied:
Good to know, I'll have to keep an eye out for that, thanks.

DanailH commented:
> Generic review
> 
> 1. It seems that we are mainly using density but for some reason, we expose it as size which is quite confusing I think. To me, size is more a measure in pixel or any units... Compact, Normal, or small seems more like a density style, and in the code, you refer to it mainly as density so I think we should unify this vocabulary and keep one terminology.
> 2. I maintain my thing for splitting the components in 2, the button and the menu, all linked by the api.
> 
> * Same way we develop the rest of the code
> * Give the flexibility to users to render the menu without the button, and outside the toolbar.
> * Allow opening the menu with api, and state

Maybe we can have a short call today to discuss this with @oliviertassinari .

For the first point - I called the prop `density` initially but @oliviertassinari made the suggestion to call it `size` instead and that's why I changed it.

For the second point - my main argument for not splitting the menu from the button is that it seems logical that if you decide to replace the button you also wouldn't want to use the menu that comes with it. Also I really think we should try and avoid adding too many public APIs, I don't think having an API to open and close a very specific menu is valuable but if we expose it then we need to support it. 

Again let's have a quick chat, maybe there is a case I'm not seeing.

DanailH commented:
Just a quick summary: I had a call with @dtassone to discuss his latest comment. We agreed to keep it as it is for the second point (splitting the menu from the button). Regarding the first point (about making the prop as `size`) he raised some concerns about the misalignment of the exposed prop name and how we call the component internally. Me personally I prefer to call the exposed prop `density`. I'm ok with keeping it as `size` but if I then need to also rename the hook to `useSize` then that would be a problem as for me if in future we decide to implement resizing of the grid (like a `textarea` for example) then the name `useSize` will be much more appropriate. What do you think @oliviertassinari ?

mbrookes commented:
Sorry, I misunderstood the purpose of this prop, not least because of its name. 🙂 

If the UI component is labelled density, its options should be "low', "medium" and "high"; and the prop to change the default selection for that component from "medium" to "low" or "high", should be called `defaultDensity`.

(The description would also need to be changed to something closer to the original, as mine proposed change doesn't correctly reflect the purpose of the prop).

However, I would question the need for this prop at all. What's the use-case?

While we're on terminology, should this be called DensityMenu rather than DensityPicker? It's at odds with the use for the date / time pickers.

oliviertassinari commented:
> misalignment of the exposed prop name and how we call the component internally.

@DanailH I definitely agree that we need to have the naming in the code consistent.

> However, I would question the need for this prop at all. What's the use-case?

@mbrookes It's meant to allow developers to control the state, in addition to the select/picker in the toolbar. Right now, it only changes the row height, we likely in the future want to change the padding in some places of the grid.

> should this be called DensityMenu rather than DensityPicker?

I wouldn't worry about it, the name is internal. I think that picker is fine. On the date picker, we are moving away from having pickers === date picker. I don't think that we should have ever introduced this confusing source.

> (or "compact", "normal", "comfortable", if you prefer)

For end-users, agree. For the developers, I disagree, I have tried to improve the case for `size` in https://github.com/mui-org/material-ui-x/pull/606#discussion_r533764519. If a majority think that `density` is better, fine by me, at least I have shared my point of view. At its core, I think that it's a tradeoff between being semantically more correct (density) vs. making it easier for developers to use the feature (size).

mbrookes commented:
> Right now, it only changes the row height

Uhh, no. It changes the default value for density. (Seems like I'm not the only one thrown by the prop name.)

>> should this be called DensityMenu rather than DensityPicker?

> I wouldn't worry about it, the name is internal.

Not true either:
https://deploy-preview-606--material-ui-x.netlify.app/components/data-grid/rendering/#density-picker-button

DanailH commented:
Ok, I gave it a thought and I propose  in order to be consistent to just call everything `size`. That means that the component will be called `SizePicker`, the hook will be `useSize` and the state will be either `contentSize` or `size` as we already have `containerSizes` and `viewportSizes` as part of the state. We will still keep the current options for size, `small`, 'medium', 'large'. What do you think about that @mbrookes @oliviertassinari @dtassone ?

oliviertassinari commented:
> As I understand it, it changes the default value for the density menu. (Seems like I'm not the only one thrown by the prop name.)

@mbrookes I'm not sure to follow. Are you saying that the prop should be called `defaultSize` and not `size`, because the behavior isn't controlled but uncontrolled?

DanailH commented:
Ok I've renamed Density to Size so there should not be any traces of `density` being used as a term in the codebase. Please review it again because I might have missed something 😅

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+{{"demo": "pages/components/data-grid/rendering/SizePickerSmallGrid.js", "bg": "inline"}}
+
+### Size picker button
```
There is an issue with the experimental feature. The documented demo doesn't work. The toolbar is hidden by default. This is related to https://github.com/mui-org/material-ui-x/pull/606#issuecomment-737130908.

DanailH replied:
Ok, yes I forgot. I enabled the toolbar for the demo

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
 
 {{"demo": "pages/components/data-grid/rendering/AutoHeightGrid.js", "bg": "inline"}}
 
+## Size
```
How about we use the two wordings at the same time so developers, when searching, e.g. with Algolia can find the content?

```suggestion
## Size/Density
```

DanailH replied:
done

oliviertassinari reviewed (changes requested):
- Could you try to rebase on `master`, the icon is buggy, I hope that #666 fixes it:

![rendering](https://user-images.githubusercontent.com/3165635/100869153-e8310000-349c-11eb-9403-a9e0ac3424e4.gif)

- I think that the end-users should see a density feature, not a size feature, https://medium.com/nextux/design-better-data-tables-4ecc99d23356. It's the wording used by Gmail.

mbrookes commented:
@DanailH 
> Ok, I gave it a thought and I propose in order to be consistent to just call everything size. [...] What do you think about that @mbrookes @oliviertassinari @dtassone ?

I think we're going about this the wrong way by aiming for API consistency, because the choice of name is fundamentally linked to UX – and UX trumps DX every time. This is a user facing feature. What do users expect this feature to be called? What makes most sense? For me "Density" > "Size".

Everything else, including the API, flows from that. If it's inconsistent with other components as a result, so be it. Developers will not be confused by the terminology if it's consistent with the user facing name.

> Ok I've renamed Density to Size

That might have been a bit premature. 🙂 

However in the Netlify preview, and running it locally I'm still seeing "Density" anyway.

@oliviertassinari 
> Are you saying that the prop should be called defaultSize and not size

I thought that because it sets the default size (that can be overridden by the user) the name should reflect that. But it also covers the case where the density button or toolbar is disabled and simply sets the density, so in that context "default" doesn't make sense. (We'll just need to word the documentation to cover both cases.)

> Should the toolbar be opt-in or opt-out?

My gut feeling is opt-in (start with the basic data grid, and add the options you want). The only argument I can think of in favor of opt-out is if the majority of grids are expected to have a toolbar, and having to add it every time would be pain.

DanailH commented:
@mbrookes Yup all but the label is called size. Only the label is called 'Density' (based on one of the comments before). Regarding the rename - I can revert it fairly easily and rename all to use density instead of size but I would like to have an agreement on this as I already did it twice 😅 . If I have to be honest my preference is also for `Density` but I thought based on Olivier's previous comments that we agreed to call it size. 

Regarding calling the prop `defaultSize` -> IMO this is not a good idea as default implies that it wont change which it can. For me personally `default` should only be used to define configuration.

Last but not least for the `toolbar` -> lets keep it as it is for now, as opt-in.

oliviertassinari commented:
> This is a user facing feature. What do users expect this feature to be called? What makes most sense? For me "Density" > "Size".

@mbrookes Are you saying the name for end-users **should match** the name for developers? I don't see why. I was hoping we could get away with "density" for end-users, hence renaming the options of the menu from, e.g. Small to -> Compact/Dense and `size` for developers. Ok, no objections to keeping them aligned.

So the conclusion is consistency between end-users and developers **>** consistency between different components. Hence, we don't only use density for the wording of end-users but also for developers. Ok, we can take note, for the future cases like this. I still find it weird to have `<DataGrid density="dense" />` and not `<DataGrid size="small" />` but I can live with it.

> That might have been a bit premature. 

Agree, I think that in cases like this, best to move to a different problem until all blockers are resolved. We are not here yet :). 

> I thought that because it sets the default size

Ok, thanks for the confirmation. We can ignore it for this pull request,. It's a flaw in the API, it impacts all the props, we have started to discuss it with @dtassone in #190. I think that we should first resolve it before implementing it here.

> if the majority of grids are expected to have a toolbar

I'm not sure the condition is true. At least, I have seen many different approaches. I think that it's great to have a toolbar for convenience but it should be easy to compose it and to not pay the bloating cost for it when not using it.

DanailH commented:
Guys I'm not sure that it makes much of a difference whether we will call it `density` or `size` in that specific case. After all this feature is basically just sugar on top of existing functionality that can can be achieved if developers pass `headerHeight` and `rowHeight` themselves.

oliviertassinari commented:
So to summarize on the latest state of the pull request:

1. [ ] Consistency between end-users and developers **is superior to** consistency between different components. We use density
2. [ ] The menu options should be capitalized. Sorry, it's what I meant by "uppercase": "Short", not "short" nor "SHORT":

<img width="170" alt="Capture d’écran 2020-12-02 à 16 06 49" src="https://user-images.githubusercontent.com/3165635/100890353-6fd83800-34b8-11eb-95be-e7e9afa3fed2.png">

3. [ ] The density icon is wrong on first render. Rebasing on `master` should be enough.

<img width="158" alt="Capture d’écran 2020-12-02 à 16 08 22" src="https://user-images.githubusercontent.com/3165635/100890500-9d24e600-34b8-11eb-99e7-836d79b34f09.png">

4. [ ] The font variant on the menu is wrong, it should be body2 not body1. I'm recommending we use font inheritance.

---

For later

4. [ ] `GridMenu` isn't implementing the keyboard navigation correctly, the focus isn't restored to the button. Will probably be fixed by killing the component once we can use Menu from the core in v5.

5. [ ] We will need to fix the controllable state API, the current approach is surprising and inconsistent #190.

mbrookes commented:
> all people will have a toolbar they never asked for

I said opt-in?

oliviertassinari commented:
> I said opt-in?

@mbrookes I agree with you, I was trying to imagine what would happen on the othe-side.

mbrookes commented:
One more item to decide for the list: Picker vs. Menu vs. Selector vs. ?

(This isn't just the internal name for the component – it's used several times in the documentation.)

DanailH commented:
Ok I've addressed the points from 1 to 4. Please check the PR for the last time hopefully. I sure hope now all the names are aligned 😅.

PS. I'll rebase from master once the issue with the infinite loop is fixed.

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/DensitySelectorHiddenGrid.tsx:
```diff
@@ -0,0 +1,17 @@
+import { DataGrid } from '@material-ui/data-grid';
+import { useDemoData } from '@material-ui/x-grid-data-generator';
+
+export default function SizePickerHiddenGrid() {
```
```suggestion
export default function DensitySelectorHiddenGrid() {
```

oliviertassinari commented on docs/src/pages/components/data-grid/rendering/DensitySelectorSmallGrid.tsx:
```diff
@@ -0,0 +1,17 @@
+import { DataGrid } from '@material-ui/data-grid';
+import { useDemoData } from '@material-ui/x-grid-data-generator';
+
+export default function SizePickerSmallGrid() {
```
```suggestion
export default function DensitySelectorSmallGrid() {
```

oliviertassinari commented on packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:
```diff
@@ -0,0 +1,116 @@
+import { densityValueSelector } from '../../hooks/features/density';
+import { GridMenu } from '../menu/GridMenu';
+
+const DensitySelectorListItemText = withStyles({
+  root: {
+    textTransform: 'capitalize',
+  },
+})(ListItemText);
```
```suggestion
```

oliviertassinari commented on packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:
```diff
@@ -0,0 +1,116 @@
+  const DensityStandardIcon = icons!.DensityStandard!;
+  const DensityComfortableIcon = icons!.DensityComfortable!;
+
+  const DensityOptions: Array<DensityOption> = [
+    {
+      icon: <DensityCompactIcon />,
+      label: DensityTypes.Compact,
+    },
+    {
+      icon: <DensityStandardIcon />,
+      label: DensityTypes.Standard,
+    },
+    {
+      icon: <DensityComfortableIcon />,
+      label: DensityTypes.Comfortable,
+    },
+  ];
```
It will need to be translated, I don't think that we can use the code constants for it.

```suggestion
  const DensityOptions: Array<DensityOption> = [
    {
      icon: <DensityCompactIcon />,
      label: 'Compact',
    },
    {
      icon: <DensityStandardIcon />,
      label: 'Standard',
    },
    {
      icon: <DensityComfortableIcon />,
      label: 'Comfortable',
    },
  ];
```

DanailH replied:
Updated

oliviertassinari commented on packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:
```diff
@@ -0,0 +1,116 @@
+
+  const handleDensitySelectorOpen = (event) => setAnchorEl(event.currentTarget);
+  const handleDensitySelectorClose = () => setAnchorEl(null);
+  const handleDensityUpdate = React.useCallback(
```
I would personally remove the using callback, we don't need the value to stay the same nor it prevents the rendering of a large React tree.
```suggestion
  const handleDensityUpdate = 
```

DanailH replied:
done

oliviertassinari commented on packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:
```diff
@@ -0,0 +1,116 @@
+      selected={option.label === densityValue}
+    >
+      <ListItemIcon>{option.icon}</ListItemIcon>
+      <DensitySelectorListItemText
+        primaryTypographyProps={{ variant: 'body2' }}
+        primary={option.label}
+      />
```
I would stick to this demo: https://material-ui.com/components/menus/#limitations, and even go one step further:

```suggestion
      {option.label}
```

DanailH replied:
Fixed

oliviertassinari commented on packages/grid/_modules_/grid/hooks/features/density/useDensity.tsx:
```diff
@@ -0,0 +1,66 @@
+import { optionsSelector } from '../../utils/useOptionsProp';
+import { DensityState } from './densityState';
+
+export const SIZE_FACTOR_SMALL = 0.7;
+export const SIZE_FACTOR_LARGE = 1.3;
```
Something like this?

```suggestion
export const COMPACT_DENSITY_FACTOR = 0.7;
export const COMFORTABLE_DENSITY_FACTOR = 1.3;
```

DanailH replied:
done

oliviertassinari commented on packages/grid/_modules_/grid/models/gridOptions.tsx:
```diff
@@ -46,9 +49,35 @@ export interface IconsOptions {
+   * Icon displayed in between two column headers that allows to resize the column header.
    */
   ColumnResize?: React.ElementType<{ className: string }>;
+  /**
+   * Icon displayed on the compact density option in the Density picker menu.
+   */
+  DensityCompact?: React.ElementType;
+  /**
+   * Icon displayed on the standard density option in the Density picker menu.
+   */
+  DensityStandard?: React.ElementType;
+  /**
+   * Icon displayed on the comfortable density option in the Density picker menu.
+   */
+  DensityComfortable?: React.ElementType;
```
```suggestion
  /**
   * Icon displayed on the compact density option in the toolbar.
   */
  DensityCompact?: React.ElementType;
  /**
   * Icon displayed on the standard density option in the toolbar.
   */
  DensityStandard?: React.ElementType;
  /**
   * Icon displayed on the comfortable density option in the toolbar.
   */
  DensityComfortable?: React.ElementType;
```

mbrookes commented on docs/src/pages/components/data-grid/rendering/DensitySelectorHiddenGrid.js:
```diff
@@ -0,0 +1,17 @@
+
+  return (
+    <div style={{ height: 300, width: '100%' }}>
+      <DataGrid {...data} hideToolbar={false} disableDensitySelector />
```
```suggestion
      <DataGrid {...data} disableDensitySelector />
```

Isn't this a NOOP? (More below...)

oliviertassinari replied:
@mbrookes It's not a noop, see #637

mbrookes replied:
Okay, thanks. Confusing.

DanailH replied:
Can I resolve this? For now we need to keep the `hideToolbar={false}` for the example

oliviertassinari replied:
I think that we should document the opposite, make the toolbar opt-in and document how to render the density toolbar.

```jsx
import * as React from 'react';
import { DataGrid, GridToolbar, GridToolbarDensitySelector } from '@material-ui/data-grid';
import { useDemoData } from '@material-ui/x-grid-data-generator';

function Header() {
  return (
    <GridToolbar>
      <GridToolbarDensitySelector />
    </GridToolbar>
  );
}

export default function DensitySelectorToolbar() {
  const { data } = useDemoData({
    dataSet: 'Commodity',
    rowLength: 10,
    maxColumns: 6,
  });

  return (
    <div style={{ height: 400, width: '100%' }}>
      <DataGrid {...data} components={{ Header }} density="compact" />
    </div>
  );
}
```

mbrookes commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+### Density prop
+
+The vertical density of the data grid can be set using the `density` prop. Regardless of its setting, the `density` prop will respect values provided using the `rowHeight` or `headerHeight` props.
```
```suggestion
The vertical density of the data grid can be set using the `density` prop. The `density` prop applies the values determined by the `rowHeight` and `headerHeight` props, if supplied.  The user can override this setting with the toolbar density selector, if provided. 
```

(Hopefully I got the meaning correct this time,  now that I better understand it. Comment if it still isn't correct, or isn't clear. 🙂)

mbrookes commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+### Density selector button
+
+The default way for a user to change the density of the data grid is by using the density selector component from the toolbar. In addition, there is a prop to hide the density selector.
```
```suggestion
The user can change the density of the data grid  by using the density selector from the toolbar. This can be hidden with the `disableDensitySelector` prop.
```

mbrookes replied:
Should this be called `hideDensitySelector` to better reflect it's function?

![image](https://user-images.githubusercontent.com/357702/101086127-4f87a680-35a8-11eb-96e2-4b478e7b8e92.png)

oliviertassinari replied:
IMHO, it would be better to drop these props and move to the opt-in toolbar + composable API, one of the last few points of #633 that isn't checked yet.

DanailH replied:
We can remove them later. For now the name matches the name of the other `toolbar` props.

mbrookes commented on docs/src/pages/components/data-grid/rendering/rendering.md:
```diff
@@ -207,6 +207,22 @@ This means that the number of rows will drive the height of the grid and consequ
+
+{{"demo": "pages/components/data-grid/rendering/DensitySelectorSmallGrid.js", "bg": "inline"}}
+
+### Density selector button
```
```suggestion
### Density selector
```

mbrookes commented on packages/grid/_modules_/grid/components/toolbar/DensitySelector.tsx:
```diff
@@ -0,0 +1,109 @@
+      {/* <DensitySelectorListItemText
+        primaryTypographyProps={{ variant: 'body2' }}
+        primary={option.label}
+      /> */}
```
Dead code

mbrookes commented:
I believe the menu should have an anchor origin of bottom left:

![image](https://user-images.githubusercontent.com/357702/101087748-94144180-35aa-11eb-8bc6-9fd341ac4531.png)

mbrookes commented:
The `rowHeight` and and `headerHeight` sections should mention how they affect density.

oliviertassinari commented:
> The rowHeight and and headerHeight sections should mention how they affect density.

@mbrookes What do you mean? "density" as a general term or "density" as the prop?

mbrookes commented:
How they affect the density options, whether set through the prop or the menu.

oliviertassinari commented:
> How they affect the density options, whether set through the prop or the menu.

@mbrookes Oh, do you mean for example that the compact option is 70% of the rowHeight value?

mbrookes commented:
Not sure it has to list the specific percentages, but it currently doesn't say that it controls the density.

oliviertassinari commented:
> Not sure it has to list the specific percentages, but it currently doesn't say that it controls the density. Maybe it doesn't need to? I don't know.

Isn't it the opposite? `density` controls the private `rowHeight` internal state.

mbrookes commented:
Since I'm talking about the documentation, I'm obviously talking about the prop, not the internal state. 🙂 

(Kind of ironic that you said it wasn't confusing†, and are now getting confused by it. 😄 )

oliviertassinari commented:
Ok, I think I see where you are getting to. You want to document that:

- row height when density is compact = `rowHeight` * 0.7
- row height when density is standard  = `rowHeight` * 1
- row height when density is confortable  = `rowHeight` * 1.3

mbrookes commented:
I'm not sure the ratios are too important, just noting the relationship between the prop and density.

DanailH commented:
Just to summarise, @mbrookes . It's not that the `rowHeight` or the `headerHeight` values affect the `density`, it's the other way around. If you mean that by setting custom `rowHeight` and `headerHeight` the amount by which the density will increase or decrease will change then yes that is correct but that is also mentioned in the documentation.

oliviertassinari reviewed (commented):
- The open animation direction is wrong, it should start from the top, not the bottom.
- The regression on `master` was fixed, we can rebase now.

DanailH commented:
@oliviertassinari so change the `hideToolbar` to `showToolbar` ?

oliviertassinari commented:
> @oliviertassinari so change the hideToolbar to showToolbar ?

@DanailH I wanted to share with Matt a long term counter proposal. It's the point 7. of https://github.com/mui-org/material-ui-x/pull/606#issuecomment-737295120. I think that we can handle once this pull request is merged and before the next release.

mbrookes commented:
> If you mean that by setting custom rowHeight and headerHeight the amount by which the density will increase or decrease will change then yes that is correct

Yes

>  but that is also mentioned in the documentation.

Yes, I know it's mentioned in the documentation, and you know that I know, because we've been discussing the best wording to use. 😀 

I was talking about the `rowHeight` or the `headerHeight` sections. There's no mention currently, but if the toolbar is opt-in, and removed from most demos, it's probably best left for the density section (hence my hesitation).

DanailH commented:
> > so change the hideToolbar to showToolbar ?
> 
> @DanailH I wanted to share with @mbrookes a counter-proposal. It's point 7. of [#606 (comment)](https://github.com/mui-org/material-ui-x/pull/606#issuecomment-737295120). I think that we can handle this once this pull request is merged and before the next release.
> 
> To answer your question, it was about the idea to remove the `hideToolbar`/`showToolbar` prop.

@oliviertassinari I really don't think squeezing this functionality as part of that PR is a good idea. This PR is already to bloated and the composable toolbar is a feature by itself which will bring even more discussions IMO. It is better to merge it as it is (maybe change the `hideToolbar` to `showToolbar`) and focus on the composable toolbar.

oliviertassinari commented:
@DanailH Agree, this is item 7. for another iteration.

oliviertassinari commented:
> The open animation direction is wrong, it should start from the top, not the bottom.

I have fixed it in #686

oliviertassinari commented on packages/storybook/src/stories/grid-toolbar.stories.tsx:
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import { XGrid } from '@material-ui/x-grid';
+import { withKnobs } from '@storybook/addon-knobs';
```
```suggestion
```

DanailH replied:
removed

oliviertassinari commented on packages/storybook/src/stories/grid-toolbar.stories.tsx:
```diff
@@ -0,0 +1,36 @@
+export default {
+  title: 'X-Grid Tests/Toolbar',
+  component: XGrid,
+  decorators: [withKnobs, withA11y],
```
```suggestion
  decorators: [withA11y],
```

DanailH replied:
done

DanailH commented:
@dtassone should I merge it ?

dtassone commented on docs/src/pages/components/data-grid/rendering/DensitySelectorHiddenGrid.tsx:
```diff
@@ -0,0 +1,17 @@
+
+  return (
+    <div style={{ height: 300, width: '100%' }}>
+      <DataGrid {...data} hideToolbar={false} disableDensitySelector />
```
` hideToolbar={false} ` is redundant, not sure the effect it will have on ppl looking at the source. But I would ask myself why did they put that.
I think we should remove them and wait until/if/when we do the refactoring. We should not think ahead of ourselves.

oliviertassinari replied:
I'm not sure that it really matters, we talked about deferring this problem to a second pull request in https://github.com/mui-org/material-ui-x/pull/606#issuecomment-739489334.

dtassone replied:
I see, but yeah normally you leave things clean because if you change your plan, and defer this refactoring as other priorities come first, then you will have this weird prop for some time in your documentation.

oliviertassinari replied:
Ok why not

oliviertassinari replied:
However, I don't think that developer will ever see this prop as the plan is to remove it before the next release.

dtassone replied:
Well we should release tomorrow or wednesday if we follow our plan...
Also I would argue that if we do showToolbar, we should consider doing showFooter...

oliviertassinari replied:
> showToolbar, we should consider doing showFooter.

I disagree, I think that https://material-ui.com/guides/api/#property-naming is more important. I also think that we don't need showToolbar not hideToolbar.
